src/app/pages/storage/snapshots/snapshot-table/snapshot-list.component.ts

```diff
@@ -0,0 +1,327 @@
+            id: row.name,
+            dataset: datasetName,
+            snapshot: snapshotName,
+            properties: row.properties,
+            name: row.name,
+          } as SnapshotListRow;
+
+          if (row.properties) {
+            snapshotListRows.push({
+              ...transformedRow,
+              used: parseInt(row.properties.used.rawvalue),
+              created: row.properties.creation.parsed.$date,
+              referenced: parseInt(row.properties.referenced.rawvalue),
+            });
+          } else {
+            snapshotListRows.push(transformedRow);
+          }
+        }
+
+        return snapshotListRows;
+      }),
+      untilDestroyed(this),
+    ).subscribe((snapshotListRows) => {
+      this.createDataSource(snapshotListRows);
+      this.cdr.markForCheck();
+    }, () => {
+      this.createDataSource();
+      this.cdr.markForCheck();
+    });
+  }
+
+  getConfirmOptions(): ConfirmOptions {
+    if (this.showExtraColumns$.value) {
+      return {
+        title: this.translate.instant(helptext.extra_cols.title_hide),
+        message: this.translate.instant(helptext.extra_cols.message_hide),
+        buttonMsg: this.translate.instant(helptext.extra_cols.button_hide),
+        hideCheckBox: true,
+      };
+    }
+
+    return {
+      title: this.translate.instant(helptext.extra_cols.title_show),
+      message: this.translate.instant(helptext.extra_cols.message_show),
+      buttonMsg: this.translate.instant(helptext.extra_cols.button_show),
+      hideCheckBox: true,
+    };
+  }
+
+  createDataSource(snapshots: SnapshotListRow[] = []): void {
+    this.dataSource = new MatTableDataSource(snapshots);
+    this.dataSource.sort = this.sort;
+  }
+
+  expandRow(row: SnapshotListRow): void {
+    this.expandedRow = this.expandedRow === row ? null : row;
+    this.cdr.markForCheck();
+  }
+
+  toggleExtraColumnsDialog(): void {
+    this.showExtraColumns$.next(!this.showExtraColumns$.value);
+    window.localStorage.setItem('snapshotXtraCols', this.showExtraColumns$.value.toString());
+    if (this.showExtraColumns$.value) {
+      this.store$.dispatch(loadSnapshots({ extra: true }));
+    }
+    this.getSnapshots();
+    this.selection.clear();
+    const slideToggleControl = this.toolbarConfig.controls.find((control) => control.name === 'extra-columns');
+    slideToggleControl.confirmOptions = this.getConfirmOptions();
+    this.cdr.markForCheck();
+  }
+
+  onAction(event: SnapshotListEvent): void {
+    switch (event.action) {
+      case 'rollback':
+        this.doRollback(event.row);
+        break;
+      case 'clone':
+        this.doClone(event.row);
+        break;
+      case 'delete':
+        this.doDelete(event.row);
+        break;
+    }
+  }
+
+  setupToolbar(): void {
+    this.settingsEvent$.pipe(
+      untilDestroyed(this),
+    ).subscribe((event: CoreEvent) => {
+      switch (event.data.event_control) {
+        case 'filter':
+          this.filterString = event.data.filter;
```

Comment:
This is probably not correct from state management perspective. ngrx state is supposed to be the source of truth, here we have two sources of truth: MatDataSource and the state. I don't have a good answer on what to do with it at the moment, but we'd need to figure this out before refactoring another table.

Reply:
Created a ticket for that - https://jira.ixsystems.com/browse/NAS-114886